runtime/compiler/trj9/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,214 @@
+
+
+int32_t TR_JProfilingRecompLoopTest::getLoopNestingLevel(int32_t nestingFactor)
+   {
+   if (nestingFactor == 1)
+      nestingFactor = 0;
+   
+   switch(nestingFactor)
+      {
+      case 0: return 0;
+      case 10: return 1;
+      case 100: return 2;
+      case 1000: return 3;
+      case 10000: return 4;
+      default: return 5;
+      }
+   }
+
+int32_t
+TR_JProfilingRecompLoopTest::perform()
+   {
+   if (comp()->getOption(TR_EnableJProfiling))
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled, run JProfiling\n");
+      }
+   else if (comp()->getProfilingMode() == JProfiling)
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has been enabled for profiling compilations, run JProfilingRecompLoopTest\n");
+      }
+   else
+      {
+      if (trace())
+         traceMsg(comp(), "JProfiling has not been enabled, skip JProfilingRecompLoopTest\n");
+      return 0;
+      }
+   RecompilationTestLocations testLocations(RecompilationTestLocationAllocator(comp()->trMemory()->currentStackRegion()));
+   TR::TreeTop *cursor = comp()->getStartTree();
+   TR::CFG *cfg = comp()->getFlowGraph();
+
+   while (cursor)
+      {
+      if (cursor->getNode()->getOpCodeValue() == TR::asynccheck && cursor->getEnclosingBlock()->getStructureOf()->getContainingLoop() != NULL)
+         {
+         int32_t loopDepth = 1;
+         TR::Block *block = cursor->getEnclosingBlock();
+         TR_RegionStructure *region = block->getParentStructureIfExists(comp()->getFlowGraph());
+         if (region)
+            region->calculateFrequencyOfExecution(&loopDepth);
+         loopDepth = getLoopNestingLevel(loopDepth);
+         testLocations.push_back(std::make_pair(cursor, loopDepth)); 
+         }
+      cursor = cursor->getNextTreeTop(); 
+      }
+   if (!testLocations.empty())
+      addRecompilationTests(comp(), testLocations);
+   return 1;
+   }
+
+
+void 
+TR_JProfilingRecompLoopTest::addRecompilationTests(TR::Compilation *comp, RecompilationTestLocations &testLocations)
+   {
+   TR_PersistentProfileInfo *profileInfo = comp->getRecompilationInfo()->findOrCreateProfileInfo();
+   TR_BlockFrequencyInfo *bfi = TR_BlockFrequencyInfo::get(profileInfo);
+   TR::CFG *cfg = comp->getFlowGraph();
+   cfg->setStructure(NULL);
+
+   for (auto testLocation = testLocations.begin(); testLocation != testLocations.end(); ++testLocation)
+      {
+      TR::TreeTop *asyncCheckTreeTop = testLocation->first;
+      TR::Node *node = asyncCheckTreeTop->getNode();
+      int32_t depth = testLocation->second;
+      if (trace())
+         traceMsg(comp, "block_%d, n%dn, depth = %d\n",asyncCheckTreeTop->getEnclosingBlock()->getNumber(), asyncCheckTreeTop->getNode()->getGlobalIndex(), depth);
+      TR_ByteCodeInfo bci = asyncCheckTreeTop->getNode()->getByteCodeInfo();
+      performTransformation(comp, "%s Add recompilation test after asyncCheck node n%dn\n", optDetailString(), node->getGlobalIndex());
```

Comment:
this needs to be in an if - performTransformation returns true if you should do the transformation and false otherwise. If it is false you shouldn't perform the transformation.